cmd/internal/server/handlers/schema_aware_serializer.go

```diff
@@ -0,0 +1,229 @@
+	for _, field := range result.Fields {
+		columns = append(columns, field.Name)
+	}
+
+	for _, row := range result.Rows {
+		record := make(map[string]*fivetransdk.ValueType)
+		for idx, val := range row {
+			if idx < len(columns) {
+
+				colName := columns[idx]
+				if selected := s.columnSelection[colName]; !selected {
+					continue
+				}
+				var (
+					fVal *fivetransdk.ValueType
+					err  error
+				)
+
+				writer, ok := s.columnWriters[colName]
+				if ok {
+					fVal, err = writer(val)
+				} else {
+					return nil, fmt.Errorf("no column writer available for %v", colName)
+				}
+
+				if err != nil {
+					return nil, errors.Wrap(err, "unable to serialize row")
+				}
+				record[colName] = fVal
+			}
+		}
+		data = append(data, record)
+	}
+
+	return data, nil
+}
+
+func NewSchemaAwareSerializer(sender LogSender, prefix string, serializeTinyIntAsBool bool, schemaList *fivetransdk.SchemaList) Serializer {
+	return &schemaAwareSerializer{
+		prefix:                 prefix,
+		sender:                 sender,
+		serializeTinyIntAsBool: serializeTinyIntAsBool,
+		schemaList:             schemaList,
+	}
+}
+
+func (l *schemaAwareSerializer) Info(msg string) {
+	l.Log(fivetransdk.LogLevel_INFO, msg)
+}
+
+func (l *schemaAwareSerializer) State(sc lib.SyncState) error {
+	state, err := json.Marshal(sc)
+	if err != nil {
+		return l.Log(fivetransdk.LogLevel_SEVERE, fmt.Sprintf("%q", err))
+	}
+	return l.sender.Send(&fivetransdk.UpdateResponse{
+		Response: &fivetransdk.UpdateResponse_Operation{
+			Operation: &fivetransdk.Operation{
+				Op: &fivetransdk.Operation_Checkpoint{
+					Checkpoint: &fivetransdk.Checkpoint{
+						StateJson: string(state),
+					},
+				},
+			},
+		},
+	})
+}
+
+func (l *schemaAwareSerializer) Log(level fivetransdk.LogLevel, s string) error {
+	return l.sender.Send(&fivetransdk.UpdateResponse{
+		Response: &fivetransdk.UpdateResponse_LogEntry{
+			LogEntry: &fivetransdk.LogEntry{
+				Message: l.prefix + " " + s,
+				Level:   level,
+			},
+		},
+	})
+}
+
+func (l *schemaAwareSerializer) Release() {
+	l.recordResponse = nil
+	l.sender = nil
+}
+
+func (l *schemaAwareSerializer) Record(result *sqltypes.Result, schema *fivetransdk.SchemaSelection, table *fivetransdk.TableSelection) error {
+	// make one response type per schema + table combination
+	// so we can avoid instantiating one object per table, and instead
+	// make one object per schema + table combo
+	if responseKey(schema, table) != l.recordResponseKey {
+		l.recordResponseKey = responseKey(schema, table)
+		l.recordResponse = &fivetransdk.UpdateResponse{
+			Response: &fivetransdk.UpdateResponse_Operation{
+				Operation: &fivetransdk.Operation{
+					Op: &fivetransdk.Operation_Record{
+						Record: &fivetransdk.Record{
+							SchemaName: &schema.SchemaName,
+							TableName:  table.TableName,
+							Type:       fivetransdk.OpType_UPSERT,
+							Data:       nil,
+						},
+					},
+				},
+			},
+		}
+		if l.serializers == nil {
```

Comment:
By doing this init in the constructor, you can avoid potential logic bugs and data races later on. Lazy init is usually more clever and problematic than it is worth.